apps/synapse-oauth-signin/src/OAuth2Form.tsx

```diff
@@ -120,37 +120,59 @@ export function OAuth2Form() {
     )
   }
 
+  /**
+   * Parse the URL search params to construct an OIDC Authorization Request
+   *
+   * @returns OIDCAuthorizationRequest if the request is valid
+   * @returns undefined if the request is invalid. `onError` will be called with an error message.
+   */
   const oidcAuthorizationRequestFromSearchParams:
     | OIDCAuthorizationRequest
     | undefined = useMemo(() => {
+    const missingParams: string[] = []
+
     const clientId = queryParams.get('client_id')
+    if (clientId == null) {
+      missingParams.push('client_id')
+    }
     const scope = queryParams.get('scope')
+    if (scope == null) {
+      missingParams.push('scope')
+    }
     const responseType = queryParams.get('response_type')
+    if (responseType == null) {
+      missingParams.push('response_type')
+    }
     const redirectUri = queryParams.get('redirect_uri')
+    if (redirectUri == null) {
+      missingParams.push('redirect_uri')
+    }
 
-    if (
-      clientId == null ||
-      scope == null ||
-      responseType == null ||
-      redirectUri == null
-    ) {
-      // We don't have the params to construct the request
+    if (missingParams.length > 0) {
+      // We don't have the params required to construct the request
+      onError(
+        new Error(
+          `Invalid request. Missing required parameter(s): ${missingParams.join(
+            ', ',
+          )}`,
+        ),
+      )
```

Comment:
Previously, the error was handled server-side. The regression occurred because I added client-side validation to prevent sending an invalid request to the server, but didn't show an error if the validation failed. Now, an error is shown if any required params are missing.